### ISSUE TITLE
Absolute path to MySQL executable

### ISSUE THREAD

**Issue by marcusatbang** [wp-cli/db-command#10]
After weeks tracking down an obscure error in our setup, we finally identified the cause of the problem.

- The db module falls back to shell for commands such as `self::run( 'mysql --no-defaults --no-auto-rehash', $assoc_args );`
- My bash profile had an alias for `mysql` that added an option `--prompt='...'` (the actual prompt was more complex, ommitted for simplicity.
- The `--no-defaults` option demands to be the _first_ option on the command line, but has a really unhelpful error message (it claims `unknown option '--no-defaults'`, which isn't true for either MySQL or MariaDB).

**Steps to reproduce:**

1. Set an alias in your bash profile such as:
`alias mysql='/usr/bin/mysql --prompt='mysql> '`
2. Open a new shell that picks up that alias
3. Call `wp db cli`

**Fix:**

On our own side, the fix is simply to remove the alias so the environment is normal.

But it also seems that WP-CLI should be calling the MySQL executable absolutely rather than relying on the individual's environment. Any code that references `mysql`, `mysqlcheck`, `mysqldump` etc should instead use `env`, eg:
```
self::run( '/usr/bin/env mysql --no-defaults --no-auto-rehash', $assoc_args );
```

**Comment by danielbachhuber** [comment]
> But it also seems that WP-CLI should be calling the MySQL executable absolutely rather than relying on the individual's environment.

If there are multiple MySQL executables installed on the system, how would WP-CLI know which to call?

Using `$PATH` to specify an executable is a common use case (e.g. when MAMP installs a second web server on your system).

**Comment by danielbachhuber** [comment]
@schlessera What are your thoughts on this?

**Comment by schlessera** [comment]
`/usr/bin/env mysql` should use the `$PATH` to find the correct `mysql` executable to use, so I don't think there's an issue with MAMP or similar stuff. Also, if there are multiple executables that match, adapting the `$PATH` is the standard way of making sure the correct one is chosen.

An alternative might be to show a warning when an alias for `mysql` has been detected, but that might break scripts that don't expect this warning.

**Comment by marcusatbang** [comment]
Thank you.

**Comment by AdsonCicilioti** [comment]
Hi @schlessera and @danielbachhuber ,
Resurrecting the topic a little, I would like your help to better understand how I can suppress or work around this problem.

I am using wpcli installed in `~/.local/bin`. Also in the same directory are two Bash Script binaries, mysql and mysqldumb with the following lines:
mysql:
```bash
docker exec mysql-lled /bin/bash -c "mysql $@"`
```
mysqldump:
```bash
docker exec mysql-lled /bin/bash -c "mysqldump $@"
```
Yes .. my mysql is a docker container with the name mysql-lled.
Basically these binaries work very well for what I need, and I did it to supply the query `/usr/bin/env mysql` that wpcli does.

But when trying to export a DB with the command:

```bash
wp db export db.sql
```
I have the output:
```bash
Usage: mysqldump [OPTIONS] database [tables]
OR mysqldump [OPTIONS] --databases [OPTIONS] DB1 [DB2 DB3 ...]
OR mysqldump [OPTIONS] --all-databases [OPTIONS]
For more options, use mysqldump --help
```

I tried to add the root user in the line of my script:
`docker exec mysql-lled /bin/bash -c" mysqldump -uroot $@"`

I run the command `wp db export db.sql` again and I get:
```bash
mysqldump: [ERROR] unknown option '--no-defaults'
```

I do not know exactly the function of this argument. Was there any way to supply this?
